pitaya-sharp/NPitaya/src/Metrics/MetricsReporter.cs

```diff
@@ -1,103 +1,37 @@
-            var name = Marshal.PtrToStringAnsi(opts.Name);
-            var prometheusReporter = RetrievePrometheus(prometheusPtr);
-            var key = BuildKey(name);
-            prometheusReporter.RegisterCounter(key);
-        }
-
-        static void RegisterHistogramFn(IntPtr prometheusPtr, MetricsOpts opts)
-        {
-            var ns = Marshal.PtrToStringAnsi(opts.MetricNamespace);
-            // TODO (felipe.rodopoulos): prometheus-net does not support subsystem label yet. We'll use a hardcoded one.
-            var name = Marshal.PtrToStringAnsi(opts.Name);
-            var prometheusReporter = RetrievePrometheus(prometheusPtr);
-            var key = BuildKey(name);
-            prometheusReporter.RegisterHistogram(key);
-        }
-
-        static void RegisterGaugeFn(IntPtr prometheusPtr, MetricsOpts opts)
-        {
-            var ns = Marshal.PtrToStringAnsi(opts.MetricNamespace);
-            // TODO (felipe.rodopoulos): prometheus-net does not support subsystem label yet. We'll use a hardcoded one.
-            var name = Marshal.PtrToStringAnsi(opts.Name);
-            var prometheusReporter = RetrievePrometheus(prometheusPtr);
-            var key = BuildKey(name);
-            prometheusReporter.RegisterGauge(key);
-        }
-
-        static void IncCounterFn(IntPtr prometheusPtr, IntPtr name, ref IntPtr labels, UInt32 labelsCount)
-        {
-            string nameStr = Marshal.PtrToStringAnsi(name);
-            var prometheus = RetrievePrometheus(prometheusPtr);
-            prometheus.IncCounter(nameStr);
-        }
-
-        static void ObserveHistFn(IntPtr prometheusPtr, IntPtr name, double value, ref IntPtr labels, UInt32 labelsCount)
-        {
-            string nameStr = Marshal.PtrToStringAnsi(name);
-            var key = BuildKey(nameStr);
-            var prometheus = RetrievePrometheus(prometheusPtr);
-            prometheus.ObserveHistogram(key, value);
-        }
-
-        static void SetGaugeFn(IntPtr prometheusPtr, IntPtr name, double value, ref IntPtr labels, UInt32 labelsCount)
+        internal MetricsReporter(MetricsConfiguration config)
         {
-            string nameStr = Marshal.PtrToStringAnsi(name);
-            var prometheus = RetrievePrometheus(prometheusPtr);
-            prometheus.SetGauge(nameStr, value);
+            _isEnabled = config.IsEnabled;
+            if (!_isEnabled)
+            {
+                return;
+            }
+            _prometheusServer = new PrometheusReporter(config);
+            _pitayaMetrics = new PitayaReporter(_prometheusServer);
         }
 
-        static void AddGaugeFn(IntPtr prometheusPtr, IntPtr name, double value, ref IntPtr labels, UInt32 labelsCount)
+        internal void Start()
         {
-            string nameStr = Marshal.PtrToStringAnsi(name);
-            Logger.Warn($"Adding gauge {nameStr} with val {value}. This method should not be used.");
-        }
+            if (!_isEnabled)
+            {
```

Comment:
I think the code is easier to maintain if in the case of disabled metrics, you don't even instance this class. Otherwise we'll need to keep checking for this boolean variable every time we make changes to the code.

Reply:
The problem is that the Rust code initialization method obligates me to send some kind of pointer in the `rawMetricsReporter`. It's a matter of having that ternary conditional on the Rust initalize constructor, or having it's managed through Config + a class that embeds all logic related to this. I prefer the latter, which seems more organized and more encapsulated.

Reply:
Rust does not obligate you to do that:
```rust 
let builder = if raw_metrics_reporter.is_null() {
    builder
} else {
    let raw_metrics_reporter = unsafe { Box::from_raw(raw_metrics_reporter) };
    builder.with_metrics_reporter(Arc::new(RwLock::new(raw_metrics_reporter)))
};
```

Did you test this and it was broken?

Reply:
Based on private discussions, we settled that it's better to the feature flag lives outside the class. Refactore.